### DISCUSSION TITLE
Proposal: Init-only methods

### DISCUSSION THREAD

**Discussion by Lehonti** [dotnet/csharplang#3578] [Language Ideas]
_**Background**: Something similar to this was proposed in the past (#3021). At that time, someone answered that it would not be possible "...without some kind of changes to the runtime..."; but with the advent of init-only properties, I would guess that any necessary changes have been / are being made (?). Furthermore, the proposal could be updated to account for said init-only properties._

---

Currently, `readonly` fields can only be assigned from the class constructor or init accessors. The reasoning behind this is sound. Sometimes we need to turn a reference (or value, in the case of structs) to some dependency into an invariant during the lifetime of an instance.

But if (say) the number of fields to assign from the constructor is very large, the initialization code becomes very long, the constructor becomes really cluttered, and there is no way to work around it. The compiler won't let me organize the initialization into neat, separate methods.

So what about this? 

``` cs
class Foo {
    
    Foo (
        IDependencyProvider dependencies
    ) {
        Initialize ( dependencies ) ;
    }
    
    // Long list of dependencies...
    readonly IDep1 _dep1;
    readonly IDep2 _dep2;
    ...
    readonly IDepN _depN;
    
    // Implicitly private, no access modifiers allowed.
    // We could, perhaps, imagine some semantics in which access modifiers
    // or virtual initialization methods would make sense and not hurt invariance
    // in unexpected ways, but let's not concern ourselves with that now...
    init void Initialize (
        IDependencyProvider dependencies
    ) {
        // Fine to assign to readonly fields
        _dep1 = dependencies.GetDep1( );
        _dep2 = dependencies.GetDep2( );
        ...
        _depN = dependencies.GetDepN();
    }
    
    init void SomeOtherInitialization ( ) { ... }
    
    ...
    
    public IBar Prop {
        get => ... ;
        init {
            ...
            SomeOtherInitialization ( ) ; // Ok
        }
    }
    
    void NonInit ( ) {
        SomeOtherInitialization ( ) ; // Error
    }
    
}
```

**Comment by HaloFour** [comment]
IMO this probably still belongs under #3021.  The language team has decided to ignore that fields are `readonly` in the context of `init` property accessors.  I don't know if the runtime is going to be modified to reconsider verifiability or if that even matters anymore.

The proposal should be updated to include `init` accessors, though.  Given that they are guarded by a specific `modreq` modifier (as of the current design) it would seem somewhat easy for the compiler to enforce that these methods can only be called from within these specific contexts, even across a type hierarchy.

**Comment by Lehonti** [comment]
@HaloFour "_...if that even matters anymore_".

Would you care to elaborate, please?

Let's say you wrote and formally verified the correctness of a program (for a security-critical application, for example) with the help of the compiler, then all of a sudden, after updating the compiler you can't do that anymore? All subsequent changes to the code will be riskier than before? Can this be fixed by (say) defining a subset of the language to which you must adhere if you want to be able to verify your program? Would it be impossible to interact with code that is not verifiable?

Sorry, my head is just overwhelmed with thoughts. I can't help but think that this is a bad move.

**Comment by quinmars** [comment]
Midori had constructor-like `init` methods: http://joeduffyblog.com/2016/11/30/15-years-of-concurrency/

**Comment by HaloFour** [comment]
@Lehonti 

This comment implies that the team does expect methods with the `modreq` to be recognized by the verifier: https://github.com/dotnet/runtime/issues/34978#issuecomment-616824545

Things have changed since then and I don't see any specific issue related to modifying the verifier, though.

**Comment by 333fred** [comment]
I'm not going to close this one: it's something that we explicitly considered in LDM and had broad support for in the meeting, we just don't have time for it in 9 with the rest of the features we're adding.

**Comment by dotMorten** [comment]
I could definitely use a common init-method, especially for when doing x-plat code, where each platform might have a different constructor, but you don't want to duplicate common initializer code.

```cs
#if WINUI
using Control = Microsoft.UI.Xaml.Controls.Control;
#elif NETFX_CORE
using Control = Windows.UI.Xaml.Controls.Control;
#elif __ANDROID__
using Control = Android.Views.ViewGroup;
#elif __IOS__
using Control = UIKit.UIView;
#elif WPF
using Control = System.Windows.Controls.Control;
#endif
namespace MyControlLibrary
{
    public class MyView : Control
    {
        private readonly string nonNullString;
#if __ANDROID__
        public MyView(Android.Content.Context context) : base(context)
        {
           CommonInit();
        }
#else
        public MyView() : base()
        {
            CommonInit();
        }
#endif
        private void CommonInit()
        {
            nonNullString = "Some string";
        }
    }
```

There's two problems with this code right now: I can't assign `nonNullString` outside a constructor, so I'd have to remove `readonly` if I want to share common init code. Secondary the nullability analyzer won't realize I guaranteed this field would be initialized.

**Comment by 333fred** [comment]
You can solve the nullability issue by using MemberNotNull, but that ofc does not fix the inability to assign.

**Comment by dotMorten** [comment]
> You can solve the nullability issue by using MemberNotNull

Sure, but it requires the developer to fulfill that promise, but there's no guarantee you are. I'd prefer not suppressing these warnings as it becomes error-prone.

**Comment by najak3d** [comment]
Bringing it to attention that I've made a similar proposal that is under discussion now, which may improve upon the proposal made here:

[https://github.com/dotnet/csharplang/discussions/5728](https://github.com/dotnet/csharplang/discussions/5728)
